packages/records/lib/db/migrations/20240327151027_create_records_table.ts

```diff
@@ -0,0 +1,104 @@
+const PARTITION_COUNT = 256;
+
+function partitionTable(i: number): string {
+    return `${TABLE}_p${i}`;
+}
+
+export async function up(knex: Knex): Promise<void> {
+    await knex.transaction(async (trx) => {
+        // TABLE
+        await trx.raw(`
+            CREATE TABLE "${TABLE}" (
+            id uuid NOT NULL,
+            external_id character varying(255) NOT NULL,
+            json jsonb,
+            data_hash character varying(255) NOT NULL,
+            connection_id integer NOT NULL,
+            model character varying(255) NOT NULL,
+            created_at timestamp with time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
+            updated_at timestamp with time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
+            deleted_at timestamp with time zone,
+            sync_id uuid,
+            sync_job_id integer,
+            CONSTRAINT ${TABLE}_connection_id_external_id_model UNIQUE (connection_id, external_id, model)
+            ) PARTITION BY HASH (connection_id, model)
+        `);
+        for (let i = 0; i < PARTITION_COUNT; i++) {
+            await trx.raw(`
+                CREATE TABLE "${partitionTable(i)}" PARTITION OF "${TABLE}"
+                FOR VALUES WITH (MODULUS ${PARTITION_COUNT}, REMAINDER ${i});
+            `);
+        }
+        // TRIGGERS
+        await trx.raw(`
+             CREATE OR REPLACE FUNCTION ${TABLE}_undelete()
+             RETURNS TRIGGER AS $$
+             BEGIN
+                 IF OLD.deleted_at IS NOT NULL AND NEW.deleted_at IS NULL THEN
+                     NEW.created_at = NOW();
+                     NEW.updated_at = NOW();
+                 END IF;
+                 RETURN NEW;
+             END;
+             $$ LANGUAGE plpgsql;
+        `);
+        await trx.raw(`
+                     CREATE TRIGGER ${TABLE}_undelete_trigger
+                     BEFORE UPDATE ON ${TABLE}
+                     FOR EACH ROW
+                     EXECUTE FUNCTION ${TABLE}_undelete();
+        `);
+        await knex.raw(`
+            CREATE OR REPLACE FUNCTION ${TABLE}_updated_at()
+            RETURNS TRIGGER AS $$
+            BEGIN
+                IF OLD.data_hash IS DISTINCT FROM NEW.data_hash THEN
+                    NEW.updated_at = NOW();
+                END IF;
+
+                RETURN NEW;
+            END;
+          $$ LANGUAGE plpgsql;
+        `);
+        await knex.raw(`
+            CREATE TRIGGER ${TABLE}_updated_at_trigger
+            BEFORE UPDATE ON ${TABLE}
+            FOR EACH ROW
+            EXECUTE FUNCTION ${TABLE}_updated_at();
+        `);
+        // INDEXES
+        await knex.schema.alterTable(TABLE, function (table) {
+            table.index(['connection_id', 'model']);
+            table.index(['connection_id', 'model', 'external_id']);
+            table.index(['connection_id', 'model', 'updated_at']);
+            table.index('sync_id');
+            table.index('sync_job_id');
+            table.unique(['connection_id', 'model', 'id']);
```

Comment:
@khaliqgant @bodinsamuel can you double check that I have not forgotten required indexes. 
Note: the `records` package doesn't support the legacy functions to fetch records.

Reply:
Looking at the current indexes, it is missing:
- id
- nango_connection_id, model, updated_at, id
- external_id

Reply:
> Looking at the current indexes, it is missing:
> 
>     * id
The idea is that id should be unique but because it is a partition table `unique constraint on partitioned table must include all partitioning columns` which is cover by `table.unique(['connection_id', 'model', 'id'])` 
> 
>     * nango_connection_id, model, updated_at, id
Fixed L76 to include `id` that I had forgotten
> 
>     * external_id
Which query would benefit from a `index(external_id)`?

Reply:
> * id

I was mostly thinking about when we query by ID not uniqueness.

> Which query would benefit from a index(external_id)?

In theory none, but it was used by the query planner in the last month to yield 500K rows